Create Orleans Quickstart (#33075)
dotnet/docs: docs/orleans/quickstarts/build-your-first-orleans-app.md

# [Visual Studio](#tab/visual-studio)

- Right click on the **OrleansURLShortener** project node in the solution explorer and select **Manage NuGet Packages**.
- In the package manager window, search for *Orleans*.
- Select the **Orleans.Server** package from the search results and select **Install**.

# [Visual Studio Code](#tab/visual-studio-code)

- In the Visual Studio code terminal, run the following command:

```dotnetcli
dotnet add package Microsoft.Orleans.Server
```

---

Open the *program.cs* file and replace the existing content with the following code:

```csharp
using Microsoft.AspNetCore.Http.Extensions;
using Orleans.Runtime;

var builder = WebApplication.CreateBuilder();

var app = builder.Build();

app.Run();
```

## Configure the silos

[Silos](/dotnet/orleans/overview#what-are-silos) are responsible for storing grains and are another core building block of Orleans. A silo can contain one or more grains; a group of silos is known as a cluster. The cluster coordinates work between silos, allowing communication with grains as though they were all available in a single process. 

> At the top of the Program class, refactor the builder code to use Orleans. The following code uses a `SiloBuilder` class to create a localhost cluster with a silo that can store grains. The `SiloBuilder` also uses the `AddMemoryGrainStorage` to configure the Orleans silos to persist grains in memory. This scenario uses local resources for development, but a production app can be configured to use highly scalable clusters and storage using services like Azure Blob Storage.

Reviewer: ```suggestion
At the top of the Program class, refactor the builder code to use Orleans. The following code uses a <xref:Orleans.Hosting.ISiloBuilder> class to create a localhost cluster with a silo that can store grains. The `ISiloBuilder` also uses the `AddMemoryGrainStorage` to configure the Orleans silos to persist grains in memory. This scenario uses local resources for development, but a production app can be configured to use highly scalable clusters and storage using services like Azure Blob Storage.
```

Also, xref to `AddMemoryGrainStorage`.

```csharp
var builder = WebApplication.CreateBuilder();

builder.Host.UseOrleans(siloBuilder =>
{
    siloBuilder.UseLocalhostClustering();
    siloBuilder.AddMemoryGrainStorage("urls");
});

var app = builder.Build();

app.Run();
```

## Create the URL shortener grain

> [Grains](/dotnet/orleans/overview#what-are-grains) are the most essential primitives and building blocks of Orleans applications. A grain is a class that inherits from the Grain base class, which manages various internal behaviors and integration points with the Orleans framework. Grains should also implement one of the interfaces listed below to define the type of grain key identifier. Each of these interfaces defines a similar contract, but marks your class with a different data type for the identifier that Orleans uses to track the grain, such as a string or integer.

Reviewer: Try to use relative links to articles, where possible.

> [Grains](/dotnet/orleans/overview#what-are-grains) are the most essential primitives and building blocks of Orleans applications. A grain is a class that inherits from the Grain base class, which manages various internal behaviors and integration points with the Orleans framework. Grains should also implement one of the interfaces listed below to define the type of grain key identifier. Each of these interfaces defines a similar contract, but marks your class with a different data type for the identifier that Orleans uses to track the grain, such as a string or integer.

Reviewer: ```suggestion
[Grains](/dotnet/orleans/overview#what-are-grains) are the most essential primitives and building blocks of Orleans applications. A grain is a class that inherits from the `Grain` base class (or `IGrainBase` interface), which manages various internal behaviors and integration points with the Orleans framework. Grains should also implement one of the interfaces listed below to define the type of grain key identifier. Each of these interfaces defines a similar contract, but marks your class with a different data type for the identifier that Orleans uses to track the grain, such as a string or integer.
```

> - IGrainWithGuidKey
> - IGrainWithIntegerKey
> - IGrainWithStringKey
> - IGrainWithGuidCompoundKey
> - IGrainWithIntegerCompoundKey

Reviewer: ```suggestion
- `IGrainWithGuidKey`
- `IGrainWithIntegerKey`
- `IGrainWithStringKey`
- `IGrainWithGuidCompoundKey`
- `IGrainWithIntegerCompoundKey`
```

Orleans grains can also use a custom interface to define their methods and properties. The URL shortener grain interface should define two methods:

- A `SetUrl` method to persist the original and shortened URL.
- A `GetUrl` method to retrieve the original URL using the shortened URL.

1. Inside of the `Program` class, append the following interface definition to the bottom of the file. Grains can implement different interfaces that .

    ```csharp
    public interface IUrlShortenerGrain : IGrainWithStringKey
    {
        Task SetUrl(string shortenedRouteSegment, string fullUrl);
        Task<string> GetUrl();
    }
    ```

1. Create a `UrlShortenerGrain` class using the following code. This class inherits from the `Grain` class provided by Orleans and implements the `IUrlShortenerGrain` interface you created. The class also uses the `IPersistentState` interface of Orleans to manage reading and writing state values for the URLs to the configured silo storage.

    ```csharp
    public class UrlShortenerGrain : Grain, IUrlShortenerGrain
    {

>         private IPersistentState<KeyValuePair<string, string>> _cache;

Reviewer: ```suggestion
        private readonly IPersistentState<KeyValuePair<string, string>> _cache;
```

    
        public UrlShortenerGrain(
            [PersistentState(
                stateName: "url",
                storageName: "urls")]
                IPersistentState<KeyValuePair<string, string>> state)
        {
            _cache = state;
        }
    
        public async Task SetUrl(string shortenedRouteSegment, string fullUrl)
        {
            _cache.State = new KeyValuePair<string, string>(shortenedRouteSegment, fullUrl);
            await _cache.WriteStateAsync();
        }
    
        public Task<string> GetUrl()
        {
            return Task.FromResult(_cache.State.Value);
        }
    }
    ```

## Create the endpoints

Next, create two endpoints to utilize the Orleans grain and silo configurations:

- A `/shorten/{*url}` endpoint to handle creating and storing a shortened version of the URL. The original, full URL is provided as a path parameter, and the shortened URL is returned to the user for later use.
- A `/go/{*shortenedUrl}` endpoint to handle redirecting users to the original URL using the shortened URL that is supplied as a parameter.

> Append the following code to the end of the `Program.cs` file:

Reviewer: ```suggestion
Append the following code to the end of the _Program.cs_ file:
```

        var resultBuilder = new UriBuilder(request.GetEncodedUrl())
        {
            Path = $"/go/{shortenedRouteSegment}"
        };
    
        return Results.Ok(resultBuilder.Uri);
    });

app.MapGet("/go/{shortenedRouteSegment}",
    async (IGrainFactory grains, string shortenedRouteSegment) =>
    {
        // Retrieve the grain using the shortened ID and redirect to the original URL        
        var shortenerGrain = grains.GetGrain<IUrlShortenerGrain>(shortenedRouteSegment);
        var url = await shortenerGrain.GetUrl();
    
        return Results.Redirect(url);
    });
```

## Test the app

The core functionality of the app should now work as expected, so now you can test your code.

1) Inside the Visual Studio Code terminal, run the `dotnet run` command again to launch the app.

    ```dotnetcli
    dotnet run
    ```

    The app should launch in the browser and display the familiar `Hello world!` text.

2) In the browser address bar, test the `shorten` endpoint by entering a URL path such as `{localhost}/shorten/https://microsoft.com`. The page should reload and provide a shortened URL. Copy the shortened URL to your clipboard.

> 3) Paste the shortened URL into the address bar and press enter. The page should reload and redirect you to [https://microsoft.com](https://microsoft.com).

Reviewer: ```suggestion
3) Paste the shortened URL into the address bar and press enter. The page should reload and redirect you to [https://microsoft.com](https://microsoft.com).

```

- `IGrainWithGuidKey`
- `IGrainWithIntegerKey`
- `IGrainWithStringKey`
- `IGrainWithGuidCompoundKey`
- `IGrainWithIntegerCompoundKey`

For this quickstart, you'll use the `IGrainWithStringKey`, since strings are a logical choice for working with URL values and short codes.

Orleans grains can also use a custom interface to define their methods and properties. The URL shortener grain interface should define two methods:

- A `SetUrl` method to persist the original and shortened URLs.
- A `GetUrl` method to retrieve the original URL using the shortened URL.

1. Append the following interface definition to the bottom of the _Program.cs_ file.

    ```csharp
    public interface IUrlShortenerGrain : IGrainWithStringKey
    {

>         Task SetUrl(string shortenedRouteSegment, string fullUrl);

Reviewer: ```suggestion
        Task SetUrl(string fullUrl);
```

        Task<string> GetUrl();
    }
    ```

1. Create a `UrlShortenerGrain` class using the following code. This class inherits from the `Grain` class provided by Orleans and implements the `IUrlShortenerGrain` interface you created. The class also uses the `IPersistentState` interface of Orleans to manage reading and writing state values for the URLs to the configured silo storage.

    ```csharp
    public class UrlShortenerGrain : Grain, IUrlShortenerGrain
    {
        private readonly IPersistentState<KeyValuePair<string, string>> _cache;
    
        public UrlShortenerGrain(
            [PersistentState(
                stateName: "url",
                storageName: "urls")]
                IPersistentState<KeyValuePair<string, string>> state)
        {
            _cache = state;
        }
    

>         public async Task SetUrl(string shortenedRouteSegment, string fullUrl)

Reviewer: ```suggestion
        public async Task SetUrl(string fullUrl)
```

        }
    }
    ```

## Create the endpoints

Next, create two endpoints to utilize the Orleans grain and silo configurations:

- A `/shorten/{*url}` endpoint to handle creating and storing a shortened version of the URL. The original, full URL is provided as a path parameter, and the shortened URL is returned to the user for later use.
- A `/go/{*shortenedUrl}` endpoint to handle redirecting users to the original URL using the shortened URL that is supplied as a parameter.

Inject the <xref:Orleans.IGrainFactory> interface into both endpoints. Grain Factories enable you to retrieve and manage references to individual grains that are stored in silos. Append the following code to the end of the _Program.cs_ file before the `app.Run()` method call:

```csharp
app.MapGet("/shorten/{*path}",
    async (IGrainFactory grains, HttpRequest request, string path) =>
    {
        // Create a unique, short ID
        var shortenedRouteSegment = Guid.NewGuid().GetHashCode().ToString("X");

        // Create and persist a grain with the shortened ID and full URL
        var shortenerGrain = grains.GetGrain<IUrlShortenerGrain>(shortenedRouteSegment);

>         await shortenerGrain.SetUrl(shortenedRouteSegment, path);

Reviewer: ```suggestion
        await shortenerGrain.SetUrl(path);
```

    });

app.MapGet("/go/{shortenedRouteSegment}",
    async (IGrainFactory grains, string shortenedRouteSegment) =>
    {
        // Retrieve the grain using the shortened ID and redirect to the original URL        
        var shortenerGrain = grains.GetGrain<IUrlShortenerGrain>(shortenedRouteSegment);
        var url = await shortenerGrain.GetUrl();
    
        return Results.Redirect(url);
    });
```

## Test the completed app

The core functionality of the app is now complete and ready to be tested. The final app code should match the following example:

```csharp
using Microsoft.AspNetCore.Http.Extensions;
using Orleans.Runtime;

var builder = WebApplication.CreateBuilder();

builder.Host.UseOrleans(siloBuilder =>
{
    siloBuilder.UseLocalhostClustering();
    siloBuilder.AddMemoryGrainStorage("urls");
});

var app = builder.Build();

app.MapGet("/", () => "Hello World!");

app.MapGet("/shorten/{*path}",
    async (IGrainFactory grains, HttpRequest request, string path) =>
    {
        var shortenedRouteSegment = Guid.NewGuid().GetHashCode().ToString("X");
        var shortenerGrain = grains.GetGrain<IUrlShortenerGrain>(shortenedRouteSegment);

>         await shortenerGrain.SetUrl(shortenedRouteSegment, path);

Reviewer: ```suggestion
        await shortenerGrain.SetUrl(path);
```

        var resultBuilder = new UriBuilder(request.GetEncodedUrl())
        {
            Path = $"/go/{shortenedRouteSegment}"
        };

        return Results.Ok(resultBuilder.Uri);
    });

app.MapGet("/go/{shortenedRouteSegment}",
    async (IGrainFactory grains, string shortenedRouteSegment) =>
    {
        var shortenerGrain = grains.GetGrain<IUrlShortenerGrain>(shortenedRouteSegment);
        var url = await shortenerGrain.GetUrl();

        return Results.Redirect(url);
    });

app.Run();

public interface IUrlShortenerGrain : IGrainWithStringKey
{

>     Task SetUrl(string shortenedRouteSegment, string fullUrl);

Reviewer: ```suggestion
    Task SetUrl(string fullUrl);
```

1. Append the following interface definition to the bottom of the _Program.cs_ file.

    ```csharp
    public interface IUrlShortenerGrain : IGrainWithStringKey
    {
        Task SetUrl(string fullUrl);
        Task<string> GetUrl();
    }
    ```

1. Create a `UrlShortenerGrain` class using the following code. This class inherits from the `Grain` class provided by Orleans and implements the `IUrlShortenerGrain` interface you created. The class also uses the `IPersistentState` interface of Orleans to manage reading and writing state values for the URLs to the configured silo storage.

    ```csharp
    public class UrlShortenerGrain : Grain, IUrlShortenerGrain
    {
        private readonly IPersistentState<KeyValuePair<string, string>> _state;
    
        public UrlShortenerGrain(
            [PersistentState(
                stateName: "url",
                storageName: "urls")]
                IPersistentState<KeyValuePair<string, string>> state)
        {
            _state = state;
        }
    
        public async Task SetUrl(string fullUrl)
        {
            _state.State = new UrlDetails(this.GetPrimaryKeyString(), fullUrl);

>             await _cache.WriteStateAsync();

Reviewer: ```suggestion
            await _state.WriteStateAsync();
```
